src/scripting/game_object.hpp

```diff
@@ -54,7 +55,7 @@
   auto object_ptr = GameObject<::OBJECT>::get_object_ptr();             \
   if (object_ptr == nullptr) {                                          \
     log_fatal << "error: script is accessing a dead object: "           \
-              << GameObject<::OBJECT>::m_uid << std::endl;              \
+              << std::endl;                                             \
     return;                                                             \
```

Comment:
**tl;dr** Regarding the last paragraph of the PR description, the presumed problem ("[...] having to search through the whole current GameObjectManager [...]") does not exist (e. g. it doesn't search), and the new proposed solution has a fatal flaw in its implementation (crash), and fixing it would not be worth it. Details below.

---------------------------

`object_ptr` here will never be null, because the pointer comes from the constructor and will never be modified. This makes the `if` dead code, but more than that, it shows the problem with using raw pointers: they don't become null when the object they point to gets destroyed.

Presuming the original code structure does indeed have a case where an object gets destroyed while a script still holds a reference to it, verifying that an object still exists is necessary. However, that can't be done with raw pointers: since the only thing you can do with them is compare with null and dereference them, that the former cannot be used as a way to detect the pointer-to object being destroyed and that the latter (dereferencing) is undefined behavior likely to result in a segfault, pointers cannot work here.

If one would insist on using pointers here, one would need to use a remove listener for each script reference to set the internal pointer to null when necessary, which is not only difficult to implement, but its implementation would take significantly more place in the source code, and the overhead that was meant to be avoided is added back.

Speaking of overhead: it may sound surprising, but **using pointers here is not an optimisation.**

The GameObjectManager holds the GameObject's in an `unordered_map`, which is implemented as a hash map. Accessing an object per ID in a hash map is [amortized constant complexity, or O(1)](https://cplusplus.com/reference/unordered_map/unordered_map/at/). That means that regardless of the amount of stuff in the map, it always takes the same amount of time to retrieve an object: Take the given ID, hash it, and access the location identified by that hash. It does not have to search through the whole current GameObjectManager. There are some edge cases which may increase complexity if a bad hash is used, hence ["amortized"](https://stackoverflow.com/questions/200384/what-is-constant-amortized-time), but since UID is based on a builtin type (size_t), we're using the default hash for that type, which we can assume is its best implementation.

Generally speaking: [beware of _premature optimisation_](https://stackify.com/premature-optimization-evil/). 1) Have we observed that accessing an object from a script takes too much time? If not, then we shouldn't waste time trying to optimize it. 2) Is the solution observably better than the original case, enough so that the increased complexity (presuming the remove listeners, needed for this to work, are added) is worth it? This should involve running benchmarks and compare the time taken by both methods. Doing this would've revealed that the complexity is not linear as assumed, but rather constant.

In short, this fixes a non-existent problem in a fundamentally flawed way, for which a fix would be way too complex. I advise keeping the original method with UIDs.